# Tarea Programación Lineal

### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables:

+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel

La tarea consiste en plantear el problemade programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

**Solución**

Se desea minimizar el costo de producción al mismo tiempo de cubrir los requerimientos nutricionales, por lo que la ecuación de optimización se vería así:

$$
100 = \min_p+\delta_p+\min_g+\delta_g+\max_f-\delta_f+\max_s-\delta_s+rem
$$

donde $p$ es proteína, $g$ grasa, $f$ fibra, $s$ sal, $rem$ el remanente y las $\delta$ la varianza de cada uno.

Así pues, sabemos también que:

* $ x_1 = 0.10p+0.08g+0.001f+0.002s$
* $ x_2 = 0.20p+0.10g+0.005f+0.005s$
* $ x_3 = 0.15p+0.11g+0.005f+0.007s$
* $ x_4 = 0.00p+0.01g+0.100f+0.002s$
* $ x_5 = 0.04p+0.01g+0.150f+0.008s$
* $ x_5 = 0.00p+0.00g+0.000f+0.000s$

donde:
  * $ x_1 = \$0.013$
  * $ x_2 = \$0.008$
  * $ x_3 = \$0.010$
  * $ x_4 = \$0.002$
  * $ x_5 = \$0.005$
  * $ x_5 = \$0.001$

  Entonces, necesito una matriz en la que se reflejen dichos valores.

  Criterio para minimizar el costo total de los ingredientes:

  $$\min_{x_1,\dots,x_6} \left( 0.013x_1 + 0.008x_2 + 0.010x_3 + 0.002x_4 + 0.005x_5 + 0.001x_6 \right)$$

Requisitos nutricionales:
1. Proteína:
$0.10x_1 + 0.20x_2 + 0.15x_3 + 0.00x_4 + 0.04x_5 + 0.00x_6 \geq 8$

* Acondiciono para utilizar:
$ -0.10x_1 - 0.20x_2 - 0.15x_3 - 0.00x_4 - 0.04x_5 - 0.00x_6 \leq -8$

2. Grasa:
$0.08x_1 + 0.10x_2 + 0.11x_3 + 0.01x_4 + 0.01x_5 + 0.00x_6 \geq 6$

* Acondiciono para utilizar:
$ -0.08x_1 - 0.10x_2 - 0.11x_3 - 0.01x_4 - 0.01x_5 - 0.00x_6 \leq -6$

3. Fibra:
$0.001x_1 + 0.005x_2 + 0.005x_3 + 0.10x_4 + 0.15x_5 + 0.00x_6 \leq 2$

4. Sal:
$ 0.002x_1 + 0.005x_2 + 0.007x_3 + 0.002x_4 + 0.008x_5 + 0.00x_6 \leq 0.4 $

5. Total:
$x_1 + x_2 + x_3 + x_4 + x_5 + x_6 = 100$



In [3]:
import numpy as np
import scipy.optimize as opt

In [7]:
c=[0.013, 0.008, 0.01, 0.002, 0.005, 0.001]
A=[[-0.1, -0.2, -0.15, 0, -0.04, 0], #prote
   [-0.08, -0.1, -0.11, -0.01, -0.01, 0], #grasa
   [0.001, 0.005, 0.005, 0.1, 0.15, 0], #fibra
   [0.002, 0.005, 0.007, 0.002, 0.008, 0], #sal
   [1,1,1,1,1,1], #total menor o igual a 100
   [-1,-1,-1,-1,-1,-1] #total mayor o igual a 100
   ]
b=[-8, -6, 2, 0.4, 100, -100]


opt.linprog(c,A,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 0.52
              x: [ 0.000e+00  6.000e+01  0.000e+00  0.000e+00  0.000e+00
                   4.000e+01]
            nit: 2
          lower:  residual: [ 0.000e+00  6.000e+01  0.000e+00  0.000e+00
                              0.000e+00  4.000e+01]
                 marginals: [ 6.400e-03  0.000e+00  1.300e-03  3.000e-04
                              3.300e-03  0.000e+00]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 4.000e+00  0.000e+00  1.700e+00  1.000e-01
                              0.000e+00  0.000e+00]
                 marginals: [-0.000e+00 -7.000e

**Respuesta:**

El costo mínimo cubriendo todos los requerimientos nutricionales se alcanza con 60 g de carne y 40 g de gel, Resultando en un costo total de $0.52 por gramo.

### Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de cuatro plantas de generación para satisfacer la demanda diaria eléctrica en cuatro ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente.


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla.

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

Para simplificar la notación use las siguientes variables:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

**Planteamiento**

El objetivo es minimizar los costos de la siguiente manera
$$
\min_{x_1,\dots,x_9} \left( 5x_1 + 2x_2 + 7x_3 + 3x_4 + 6x_5 + 6x_6 + 6x_7 + 1x_8 + 2x_9 \right)
$$

Restricciones:

Capacidad de las plantas:
- Planta 1: $
x_1 + x_2 + x_3 \leq 80$
- Planta 2: $
x_4 + x_5 + x_6 \leq 40$
- Planta 3: $
x_7 + x_8 + x_9 \leq 60$
Demanda de las ciudades:
- Guadalajara: $
x_1 + x_4 + x_7 = 70 $
- León: $
x_2 + x_5 + x_8 = 40$
- Morelia: $
x_3 + x_6 + x_9 = 70$

A cada una de las demandas para utilizarlas como igualdad, voy a convertirlas en dos ecuaciones de menor o igual, como por ejemplo:
* Guadalajara:
  - $x_1 + x_4 + x_7 \leq 70$
  - $-x_1 - x_4 - x_7 \leq -70$



In [10]:
#costos:
c = [5, 2, 7, 3, 6, 6, 6, 1, 2]

A= [[1, 1, 1, 0, 0, 0, 0, 0, 0],  # Planta 1
    [0, 0, 0, 1, 1, 1, 0, 0, 0],  # Planta 2
    [0, 0, 0, 0, 0, 0, 1, 1, 1],   # Planta 3
    [1, 0, 0, 1, 0, 0, 1, 0, 0],   # Guadalajara ≤ 70
    [0, 1, 0, 0, 1, 0, 0, 1, 0],   # León ≤ 40
    [0, 0, 1, 0, 0, 1, 0, 0, 1],   # Morelia ≤ 70
    [-1, 0, 0, -1, 0, 0, -1, 0, 0],# Guadalajara ≥ 70
    [0, -1, 0, 0, -1, 0, 0, -1, 0],# León ≥ 40
    [0, 0, -1, 0, 0, -1, 0, 0, -1] # Morelia ≥ 70
]
b= [80, 40, 60, 70, 40, 70, -70, -40, -70]
opt.linprog(c,A,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 540.0
              x: [ 3.000e+01  4.000e+01  1.000e+01  4.000e+01  0.000e+00
                   0.000e+00  0.000e+00  0.000e+00  6.000e+01]
            nit: 6
          lower:  residual: [ 3.000e+01  4.000e+01  1.000e+01  4.000e+01
                              0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              6.000e+01]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              6.000e+00  1.000e+00  6.000e+00  4.000e+00
                              0.000e+00]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf        inf        inf
                                    inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00  0.000e+00  0.0

**Respuesta**
El costo mínimo va a ser de $540 millones de pesos para satisfacer la demanda de las tres ciudades, desde las tres plantas.

La planta 1 enviará a Gdl 30 millones de KW, a León 40 y a Morelia 10.
La planta 2 solamente 40 a Gdl, y la planta 3 únicamente 60 a Morelia.

### Fábrica de focos

Una fábrica produce focos incandesentes y los vende a 4.5 u.m., además de focos ahorradores que vende a 6 u.m. cada uno. La producción está limitada a 400 focos incandesentes y 300 focos ahorradores al día, además de no poder producir más de 500 focos en total.

Si la fábrica vende toda su producción, determina cuántos focos incandesentes y cuántos focos ahorradores debe producir para obtener los máximos ingresos posibles y cuáles serían éstos.

**Optimización**

Defino las variables:
* $x_1$= Focos incandescentes producidos
* $x_2$= Focos ahorradores producidos

mi objetivo es:
$$\max_{x_1,x_2} \left( 4.5x_1 + 6x_2 \right)$$

y para usar linpro necesito minimizar, entonces:
$$\min_{x_1,x_2} \left( -4.5x_1 - 6x_2 \right)$$

mis restricciones son:
- Incandescentes:
$x_1 \leq 400$
- Ahorradores:
$x_2 \leq 300$
- Total:
$x_1 + x_2 \leq 500$





In [11]:
# Función objetivo
c = [-4.5, -6]

# Restricciones
A = [
    [1, 0],     # x1 ≤ 400
    [0, 1],     # x2 ≤ 300
    [1, 1]      # x1 + x2 ≤ 500
]
b = [400, 300, 500]
opt.linprog(c,A,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: -2700.0
              x: [ 2.000e+02  3.000e+02]
            nit: 1
          lower:  residual: [ 2.000e+02  3.000e+02]
                 marginals: [ 0.000e+00  0.000e+00]
          upper:  residual: [       inf        inf]
                 marginals: [ 0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 2.000e+02  0.000e+00  0.000e+00]
                 marginals: [-0.000e+00 -1.500e+00 -4.500e+00]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0

**Resultado**

La fábrica debe producir 200 focos incandescentes y 300 ahorradores, y para obtener un máximo de ingreso de 2700 u.m. debe usar toda su capacidad de producción al día.

### Problema de compra al mayoreo
Un frutero necesita 16 cajas de naranjas, 5 de plátanos y 20 de manzanas. Dos mayoristas pueden suministrarle para satisfacer sus necesidades, pero sólo venden la fruta en contenedores completos. El mayorista A envía en cada contenedor 8 cajas de naranjas, 1 de plátanos y 2 de manzanas. El mayorista B envía en cada contenedor 2 cajas de naranjas, una de plátanos y 7 de manzanas. Sabiendo que cada contenedor del mayorista A tiene un costo de 150 u.m. y el contenedor del mayorista B de 300 u.m., calcular cuántos contenedores habrá de comprar a cada mayorista, con objeto de reducir al mínimo el costo de lo solicitado.

**Planteamiento**

* $x_1$= no. contenedores comprados a A
* $x_2$= no. contenedores comprados a B

Objetivo:
$$ \min_{x_1,x_2} \left( 150x_1 + 300x_2 \right) $$

Restricciones:
- Naranjas:
$8x_1 + 2x_2 \geq 16$

- Plátanos:
$1x_1 + 1x_2 \geq 5$

- Manzanas:
$2x_1 + 7x_2 \geq 20$

** Recuerdo cambiar signos para geq

In [12]:
c = [150, 300]
A = [
    [-8, -2],   # Naranjas
    [-1, -1],   # Plátanos
    [-2, -7]    # Manzanas
]
b = [-16, -5, -20]
opt.linprog(c,A,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 1050.0
              x: [ 3.000e+00  2.000e+00]
            nit: 3
          lower:  residual: [ 3.000e+00  2.000e+00]
                 marginals: [ 0.000e+00  0.000e+00]
          upper:  residual: [       inf        inf]
                 marginals: [ 0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 1.200e+01  0.000e+00  0.000e+00]
                 marginals: [-0.000e+00 -9.000e+01 -3.000e+01]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0

**Respuesta**

Para cumplir con los requerimientos, el costo mínimo de la compra es de 1050 u.m.

Con 3 contenedores comprados a A y 2 a B. Lo que resulta en 28 cajas de naranjas, 5 de plátanos y 20 de manzanas.